File: Week6/sentimental-credit/credit.py
# TODO
from cs50 import get_int
from cs50 import get_string

while True:
    cardnum = get_string("Number: ")
    if cardnum.isdigit():
        break


validcard = 0
validcardstr = ""
validcard2 = 0
valid = False
off = "0"

if len(cardnum) % 2 == 0:
    for i in range(0, len(cardnum), 2):
        validcardstr += str(int(cardnum[i]) * 2)
        if i < len(cardnum) - 1:
            validcard2 += int(cardnum[i + 1])
else:
    off += cardnum
    for i in range(0, len(off), 2):
        validcardstr += str(int(off[i]) * 2)
        if i < len(off) - 1:
            validcard2 += int(off[i + 1])
for c in validcardstr:
    validcard += int(c)
validcard += validcard2
validcardstr = str(validcard)
if validcardstr[len(validcardstr) - 1] == "0":
    valid = True
amex = ["34", "37"]
mc = ["51", "52", "53", "54", "55"]
visa = ["4"]

if valid:
    if len(cardnum) == 15:
        if cardnum[0:2] in amex:
            print("AMEX\n")
        else:
            print("INVALID\n")
    elif len(cardnum) == 16:
        print(cardnum[0])
        if cardnum[0:2] in mc:
            print("MASTERCARD\n")
        elif cardnum[0] in visa:
            print("VISA\n")
        else:
            print("INVALID\n")
    elif len(cardnum) == 13:
        if cardnum[0] in visa:
            print("VISA\n")
        else:
            print("INVALID\n")
    else:
        print("INVALID\n")
else:
    print("INVALID\n")

File: Week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool asking = true;

void invalid(void)
{
    printf("INVALID\n");
}

int main(void)
{
    //Initializing Variables
    long cardnum = get_long("Card Number: ");
    long digits = 0;
    long digitcheck = cardnum;
    int checksum = 0;
    long x = 0;
    long dig;
    long dig2;
    long newnum = cardnum;

    //Counting digits
    while (digitcheck > 0)
    {
        digitcheck = digitcheck / 10;
        digits++;
        //printf("Digits: %li\n", digits);
    }

    long p = pow(10, digits);

    //Finding values and using Luhn's Algorithm
    while (x < digits)
    {
        dig = newnum % 10;
        newnum = newnum / 10;
        dig2 = newnum % 10;
        newnum = newnum / 10;
        //printf("digit %li: %li\n",x,dig2);
        x += 2;
        if (dig2 * 2 >= 10)
        {
            checksum += dig2 * 2 % 10 + 1;
        }
        else
        {
            checksum += dig2 * 2;
        }
        checksum += dig;
    }
    float startingnums = floor(cardnum / pow(10, digits - 2));
    //printf("sum: %i\n",checksum);
    //printf("first digits: %f\n", startingnums );

    //Starting Verification, If numbers are valid for a card.
    if (checksum % 10 == 0)
    {
        //printf("PASSED");
        if (startingnums == 34 || startingnums == 37)
        {
            if (digits == 15)
            {
                printf("AMEX\n");
            }
            else
            {
                invalid();
            }
        }

        else if (startingnums == 51 || startingnums == 52 || startingnums == 53 || startingnums == 54 || startingnums == 55)
        {
            if (digits == 16)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                invalid();
            }
        }

        else if (floor(startingnums / 10) == 4)
        {
            if (digits == 16 || digits == 13)
            {
                printf("VISA\n");
            }
            else
            {
                invalid();
            }
        }
        else
        {
            invalid();
        }
    }
    else
    {
        invalid();
    }
}